include/spinlock.h

```diff
@@ -61,6 +61,11 @@ struct k_spinlock {
 	 */
 	size_t thread_cpu;
 #endif
+
+#if !defined(CONFIG_SMP) && !defined(SPIN_VALIDATE)
+	/* Empty struct has size 0 in C, size 1 in C++ */
+	char dummy;
```

Comment:
No, this isn't right.  We need the optimization provided by the empty struct, otherwise the generated code needs to pass extra pointers around when using spinlocks, when what is desired is that they turn naturally into (argument-free) z_arch_irq_lock() calls.

If you want to remove the use of an empty struct, we need to do more work and remove the spinlock structs from the API arguments (when !SMP) and  enclosing struct declarations entirely.

If you just want C++ code to compile in uniprocessor mode, it's probably enough to put this whole file in a `#ifndef __cplusplus` and replace the handful of functions declared with simple declarations.  The spinlock API is not part of our public API and doesn't need to actually be used from C++ code, you just want it not to break when pulled in as a transitive header, right?

Reply:
The __cplusplus guard is not right IMHO, we have to get rid of empty struct. I've a patch that enclosure this struct with ifdef but looked bad to me due the amount of ifdef. I can send it or @andyross if you have an alternative from the top of you head, let me know that I change my patch.
But again, we have to remove empty structs and not workaround it.

Reply:
btw, add an element just to satisfy it is *not* right too.

Reply:
Put the whole file in a #ifndef __cplusplus isn't a valid fix, because the Zephyr kernel can be compiled as C library, and used after by C++ code.

Reply:
my fix is not ifdefing the whole file but only this struct, that is why this patch is nasty. This is not only a c++ problem btw, this violates C standard as well.

Reply:
Hello @andyross ,
I take some time to analyze z_arch_irq_lock() calls and I haved looking for the impact of the dummy member in the structure.
With default optimization value "-Os", there is no difference beetween the two version.
Compiler generates exactly the same code for calling z_arch_irq_lock(), in both cases.

So, add a "dummy" field to an empty structure doesn't have any impact to the performances.